src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -322,13 +322,35 @@ private static string GenerateUpgradeReport(NuGetProject nuGetProject, string ba
             bool acceptedLicense = true;
 
             List<string> removedPackages = null;
+            HashSet<Tuple<string, string>> existingPackages = new HashSet<Tuple<string, string>>();
             List<Tuple<string,string>> addedPackages = null;
             List<Tuple<string,string>> updatedPackagesOld = null;
             List<Tuple<string,string>> updatedPackagesNew = null;
 
             // Enable granular level telemetry events for nuget ui operation
             uiService.ProjectContext.OperationId = Guid.NewGuid();
 
+            try
+            {
+                // collect the install state of the existing packages
+                foreach (var project in uiService.Projects)
```

Comment:
Please use a stopwatch around the foreach loop, maybe even around the whole try-catch block, and add it as a measure, so we can see if this package collection is impacting customer experience.

Reply:
Done.